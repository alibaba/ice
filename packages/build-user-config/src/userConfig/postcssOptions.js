const checkPostcssLoader = (config, ruleName) =>
  config.module.rules.has(ruleName) &&
  config.module.rule(ruleName).uses.has('postcss-loader');

module.exports = (config, postcssOptions, context) => {
  if (!postcssOptions) {
    return;
  }

  const customPostcssOptions = { ...postcssOptions };

  const styleRules = [
    'css',
    'css-module',
    'css-global',
    'scss',
    'scss-module',
    'scss-global',
    'less',
    'less-module',
    'less-global',
  ];
  const { rootDir = '' } = context || {};
  let finalPostcssOptions = {};

  // get default postcss config
  if (checkPostcssLoader(config, 'css')) {
    const builtInOptions =
      config.module.rule('css').use('postcss-loader').get('options') || {};
    const { config: loaderConfigOptions } = builtInOptions;
    if (loaderConfigOptions && loaderConfigOptions.path) {
      try {
        const postcssFile = `${loaderConfigOptions.path}/defaultPostcssConfig`;
        finalPostcssOptions = {
          postcssOptions:
            (loaderConfigOptions.ctx
              ? // eslint-disable-next-line
                require(postcssFile)(loaderConfigOptions.ctx)
              : // eslint-disable-next-line
                require(postcssFile)) || {},
        };
      } catch (err) {
        console.log('[Error] fail to load default postcss config');
      }
    } else {
      finalPostcssOptions = builtInOptions;
    }

    // get plugins in different versions of postcss-loader
    let builtInPlugins = [];
    if (!finalPostcssOptions.postcssOptions) {
      // get plugins in v3
      builtInPlugins = [...(finalPostcssOptions.plugins || [])];
    } else {
      // get plugins in v5
      builtInPlugins = [...(finalPostcssOptions.postcssOptions.plugins || [])];
    }

    // merge plugins
    const finalPlugins = [...builtInPlugins];
    Object.keys(customPostcssOptions.plugins || {}).forEach((pluginName) => {
      let pluginOptions = {};
      const targetIndex = builtInPlugins.findIndex((pluginConfig) => {
        const [name, options] = Array.isArray(pluginConfig)
          ? pluginConfig
          : [pluginConfig];
        if (options) {
          pluginOptions = options;
        }
        return typeof name === 'string' && name.indexOf(pluginName) > -1;
      });
      const options = customPostcssOptions.plugins[pluginName];
      if (targetIndex > -1) {
        if (options === false) {
          // delete builtIn plugin
          finalPlugins.splice(targetIndex, 1);
        } else {
          // shallow merge for options
          const mergedOptions = { ...pluginOptions, ...options };
          finalPlugins.splice(targetIndex, 1, [pluginName, mergedOptions]);
        }
      } else {
        finalPlugins.push([pluginName, options]);
      }
    });
    delete customPostcssOptions.plugins;

    const postcssPlugins = finalPlugins.map((pluginInfo) => {
      const [name, options] = Array.isArray(pluginInfo)
        ? pluginInfo
        : [pluginInfo];
      if (typeof name === 'string') {
        const resolvePath = require.resolve(name, { paths: [rootDir] });
        // eslint-disable-next-line
        return require(resolvePath)(options);
      } else {
        return pluginInfo;
      }
    });

    if (!finalPostcssOptions.postcssOptions) {
      // set final postcss options in v3
      finalPostcssOptions = {
        ...finalPostcssOptions,
        ...customPostcssOptions,
        plugins: [...postcssPlugins],
      };
    } else {
      // set default value
      customPostcssOptions.postcssOptions =
        customPostcssOptions.postcssOptions || {};

      // modify option name `exec` to `execute` in v5
      if (customPostcssOptions.exec) {
        customPostcssOptions.execute = customPostcssOptions.exec;
        delete customPostcssOptions.exec;
      }

      const shouldMoveToPostcssOptionsKeys = [
        'parser',
        'syntax',
        'stringifier',
        'from',
        'to',
        'map',
      ];

      // move options to postcssOptions
      Object.keys(customPostcssOptions || {}).forEach((optionName) => {
        if (shouldMoveToPostcssOptionsKeys.includes(optionName)) {
          customPostcssOptions.postcssOptions[optionName] =
            customPostcssOptions[optionName];
          delete customPostcssOptions[optionName];
        }
      });

      // set final postcss options in v5
      finalPostcssOptions = {
        ...finalPostcssOptions,
        ...customPostcssOptions,
        postcssOptions: {
          ...finalPostcssOptions.postcssOptions,
          ...customPostcssOptions.postcssOptions,
          plugins: [...postcssPlugins],
        },
      };
    }

    // modify css rules
    styleRules.forEach((ruleName) => {
      if (checkPostcssLoader(config, ruleName)) {
        config.module
          .rule(ruleName)
          .use('postcss-loader')
          .tap(() => {
            // merge postcss-loader options
            return finalPostcssOptions;
          });
      }
    });
  }
};
