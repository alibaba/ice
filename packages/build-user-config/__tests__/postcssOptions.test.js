import Config from 'webpack-chain';
import postcssOptions from '../src/userConfig/postcssOptions';

let config = null;
let consoleWarn = console.warn;

// silent warn of postcss plugin
beforeAll(() => {
  console.warn = jest.fn();
});

afterAll(() => {
  console.warn = consoleWarn;
});

describe('merge postcss options when postcss-loader is v3', () => {
  beforeEach(() => {
    config = new Config();

    config.module
      .rule('css')
      .use('postcss-loader')
      .loader(require.resolve('postcss-loader'))
      .options({
        sourceMap: true,
        plugins: [
          'postcss-import',
          [
            'postcss-preset-env',
            {
              autoprefixer: {
                flexbox: 'no-2009',
              },
              stage: 3,
            },
          ],
          'postcss-plugin-rpx2vw',
        ],
      });
  });

  afterEach(() => {
    config = null;
  });

  it('should convert default options in rax correctly', () => {
    const outerOptions = {};
    postcssOptions(config, outerOptions);

    const { options: innerOptions } = config.module
      .rule('css')
      .use('postcss-loader')
      // @ts-ignore
      .toConfig();

    expect(innerOptions.sourceMap).toEqual(true);
    expect(innerOptions.plugins).toHaveLength(3);
    expect(innerOptions.postcssOptions).toBeUndefined();

    expect(innerOptions).toMatchSnapshot();
  });

  // see: https://github.com/webpack-contrib/postcss-loader/tree/v3.0.0#options
  it('should convert root options correctly', () => {
    const outerOptions = {
      exec: true,
      sourceMap: false,
      parser: 'sugarss',
      syntax: 'sugarss',
      stringifier: 'midas',
      config: {},
      plugins: [],
    };
    postcssOptions(config, outerOptions);

    const { options: innerOptions } = config.module
      .rule('css')
      .use('postcss-loader')
      // @ts-ignore
      .toConfig();

    expect(innerOptions.exec).toEqual(true);
    expect(innerOptions.sourceMap).toEqual(false);
    expect(innerOptions.parser).toEqual('sugarss');
    expect(innerOptions.syntax).toEqual('sugarss');
    expect(innerOptions.stringifier).toEqual('midas');
    expect(innerOptions.config).toMatchObject({});
    expect(innerOptions.plugins).toHaveLength(3);

    expect(innerOptions).toMatchSnapshot();
  });

  it('should merge plugin correctly', () => {
    const outerOptions = {
      plugins: {
        'postcss-plugin-rpx2vw': false,
      },
    };
    postcssOptions(config, outerOptions);

    const { options: innerOptions } = config.module
      .rule('css')
      .use('postcss-loader')
      // @ts-ignore
      .toConfig();

    expect(innerOptions.sourceMap).toEqual(true);
    expect(innerOptions.plugins).toHaveLength(2);

    expect(innerOptions).toMatchSnapshot();
  });
});

describe('merge postcss options when postcss-loader is v5', () => {
  beforeEach(() => {
    config = new Config();

    config.module
      .rule('css')
      .use('postcss-loader')
      .loader(require.resolve('postcss-loader'))
      .options({
        sourceMap: true,
        postcssOptions: {
          plugins: [
            'postcss-import',
            [
              'postcss-preset-env',
              {
                autoprefixer: {
                  flexbox: 'no-2009',
                },
                stage: 3,
              },
            ],
            'postcss-plugin-rpx2vw',
          ],
        },
      });
  });

  afterEach(() => {
    config = null;
  });

  it('should convert default options in rax correctly', () => {
    const outerOptions = {};
    postcssOptions(config, outerOptions);

    const { options: innerOptions } = config.module
      .rule('css')
      .use('postcss-loader')
      // @ts-ignore
      .toConfig();

    expect(innerOptions.sourceMap).toEqual(true);
    expect(innerOptions.plugins).toBeUndefined();
    expect(innerOptions.postcssOptions).toBeDefined();
    expect(innerOptions.postcssOptions.plugins).toHaveLength(3);

    expect(innerOptions).toMatchSnapshot();
  });

  // see: https://github.com/webpack-contrib/postcss-loader/tree/v5.3.0#options
  it('should convert root options correctly', () => {
    const outerOptions = {
      exec: true,
      sourceMap: false,
      parser: 'sugarss',
      syntax: 'sugarss',
      stringifier: 'midas',
      plugins: [],
      map: false,
    };

    postcssOptions(config, outerOptions);

    const { options: innerOptions } = config.module
      .rule('css')
      .use('postcss-loader')
      // @ts-ignore
      .toConfig();

    expect(innerOptions.execute).toEqual(true);
    expect(innerOptions.exec).toBeUndefined();
    expect(innerOptions.sourceMap).toEqual(false);
    expect(innerOptions.postcssOptions.parser).toEqual('sugarss');
    expect(innerOptions.postcssOptions.syntax).toEqual('sugarss');
    expect(innerOptions.postcssOptions.stringifier).toEqual('midas');
    expect(innerOptions.postcssOptions.plugins).toHaveLength(3);
    expect(innerOptions.postcssOptions.map).toEqual(false);

    expect(innerOptions).toMatchSnapshot();
  });

  it('should merge plugin correctly', () => {
    const outerOptions = {
      plugins: {
        'postcss-plugin-rpx2vw': false,
      },
    };
    postcssOptions(config, outerOptions);

    const { options: innerOptions } = config.module
      .rule('css')
      .use('postcss-loader')
      // @ts-ignore
      .toConfig();

    expect(innerOptions.sourceMap).toEqual(true);
    expect(innerOptions.postcssOptions.plugins).toHaveLength(2);

    expect(innerOptions).toMatchSnapshot();
  });
});
