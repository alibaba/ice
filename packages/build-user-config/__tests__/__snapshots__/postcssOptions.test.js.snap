// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge postcss options when postcss-loader is v3 should convert default options in rax correctly 1`] = `
Object {
  "plugins": Array [
    [Function],
    [Function],
    [Function],
  ],
  "sourceMap": true,
}
`;

exports[`merge postcss options when postcss-loader is v3 should convert root options correctly 1`] = `
Object {
  "config": Object {},
  "exec": true,
  "parser": "sugarss",
  "plugins": Array [
    [Function],
    [Function],
    [Function],
  ],
  "sourceMap": false,
  "stringifier": "midas",
  "syntax": "sugarss",
}
`;

exports[`merge postcss options when postcss-loader is v3 should merge plugin correctly 1`] = `
Object {
  "plugins": Array [
    [Function],
    [Function],
  ],
  "sourceMap": true,
}
`;

exports[`merge postcss options when postcss-loader is v5 should convert default options in rax correctly 1`] = `
Object {
  "postcssOptions": Object {
    "plugins": Array [
      [Function],
      [Function],
      [Function],
    ],
  },
  "sourceMap": true,
}
`;

exports[`merge postcss options when postcss-loader is v5 should convert root options correctly 1`] = `
Object {
  "execute": true,
  "postcssOptions": Object {
    "map": false,
    "parser": "sugarss",
    "plugins": Array [
      [Function],
      [Function],
      [Function],
    ],
    "stringifier": "midas",
    "syntax": "sugarss",
  },
  "sourceMap": false,
}
`;

exports[`merge postcss options when postcss-loader is v5 should merge plugin correctly 1`] = `
Object {
  "postcssOptions": Object {
    "plugins": Array [
      [Function],
      [Function],
    ],
  },
  "sourceMap": true,
}
`;
